### PR TITLE
API Add experimental getResource() to Module

### PR DESCRIPTION
**Pull request by tractorcow**
Partial implementation of https://github.com/silverstripe/silverstripe-framework/issues/6243

Also required for https://github.com/silverstripe/silverstripe-framework/issues/6735

### CODE REVIEW

**Comment by tractorcow** [comment]
Rebased.

**Comment by sminnee** [comment]
Looks good. Presumably if we updated Requirements to include assets through this method (it might need a new method signature) we could push modules into `vendor/`?

File: src/Core/Manifest/Module.php [hunk 1]
```diff
@@ -148,4 +149,33 @@ protected function loadComposer()
...
+     * This method makes no distinction between public / local resources,
+     * which may change in the near future.
+     *
+     * @internal Experimental API and may change
+     * @param string $path File or directory path relative to module directory
+     * @return string Path relative to base directory
+     */
+    public function getResource($path)
+    {
+        return File::join_paths($this->getRelativePath(), $path);
```
Comment (sminnee) [inline src/Core/Manifest/Module.php]: Let's remove the dependence on File, it's way bigger than is needed here.

File: src/Core/Manifest/Module.php [hunk 2]
```diff
@@ -148,4 +149,33 @@ protected function loadComposer()
...
+     * Directories included
+     *
+     * This method makes no distinction between public / local resources,
+     * which may change in the near future.
+     *
+     * @internal Experimental API and may change
+     * @param string $path File or directory path relative to module directory
+     * @return string Path relative to base directory
+     */
+    public function getResource($path)
```
Comment (sminnee) [inline src/Core/Manifest/Module.php]: Would suggest `getResourcePath()`?

File: src/Core/Manifest/Module.php (hunk 1 above)
Comment (chillu) [inline src/Core/Manifest/Module.php]: OK, using `rtrim()` instead

Comment (sminnee) [inline src/Core/Manifest/Module.php]: Maybe we could transpile `left-pad` for this?

File: src/Core/Manifest/Module.php [hunk 3]
```diff
@@ -148,4 +148,35 @@ protected function loadComposer()
...
+     *
+     * @internal Experimental API and may change
+     * @param string $path File or directory path relative to module directory
+     * @return string Path relative to base directory
+     */
+    public function getResourcePath($path)
+    {
+        $base = rtrim($this->getRelativePath(), '/\\');
+        $path = rtrim($path, '/\\');
+        return "{$base}/{$path}";
```
Comment (tractorcow) [inline src/Core/Manifest/Module.php]: ltrim this please; Not accounting for case where `$base` is empty string.

Comment (dhensby) [inline src/Core/Manifest/Module.php]: Who's going to PR this?

Comment (chillu) [inline src/Core/Manifest/Module.php]: https://github.com/silverstripe/silverstripe-framework/pull/6750

**Pull request merged**